src/core/internal/hash.d

```diff
@@ -30,6 +30,86 @@ private enum floatCoalesceNaNs = true;
         assert(hashOf(double.nan) == hashOf(-double.nan)); // Same hash for different NaN.
 }
 
+private template hasCallableToHash(T)
+{
+    static if (is(typeof(((T* x) => (*x).toHash())(null)) HashType))
+        enum hasCallableToHash = __traits(isIntegral, HashType) && is(HashType : size_t);
+    else
+        enum hasCallableToHash = false;
+}
+
+@nogc nothrow pure @safe unittest
+{
+    static struct S { size_t toHash() { return 4; } }
+    assert(hasCallableToHash!S);
+    assert(!hasCallableToHash!(shared const S));
+}
+
+/+
+Is it valid to calculate a hash code for T based on the bits of its
+representation? Always false for classes, interfaces, dynamic arrays,
+and associative arrays.
++/
+private template canBitwiseHash(T)
+{
+    static if (is(T EType == enum))
+        enum canBitwiseHash = .canBitwiseHash!EType;
+    else static if (is(T == class) || is(T == interface))
+        enum canBitwiseHash = false;
+    else static if (is(T == struct))
```

Comment:
@ZombineDev Updated to check for `isNested`. Would you be okay with leaving the padding issue to be addressed by a future PR?

>IIRC, according to the spec, the contents of padding gaps are treated like uninitialized variables

[Apparently at one point they were defined to be zero](https://forum.dlang.org/post/hn11oh$1usk$1@digitalmars.com
):
>**Walter Bright wrote on March 07, 2010:**
>The "holes" are defined to be filled with 0, and are when initialized by the compiler. This is specifically so that memcmp can be done to compare the struct contents.

[But the current specification says their contents are undefined](https://dlang.org/spec/struct.html#struct_layout):
>The padding data can be accessed, but its contents are undefined.

Reply:
> Would you be okay with leaving the padding issue to be addressed by a future PR?

Sure, but can you leave a comment along the lines of: "Note: according to the spec (as of https://github.com/dlang/dlang.org/commit/d66eff16491b0664c0fc00ba80a7aa291703f1f2) the contents of unnamed paddings between fields is undefined. Currently this hashing implementation assumes that the padding contents (if any) for all instances of `T` are the same."?

Reply:
Sure, where should I put it?